### ISSUE TITLE
cscope/gnu global support

### ISSUE THREAD

**Issue by badeip** [ludovicchabant/vim-gutentags#15]
It would be nice if you can support **GNU Global** or **Cscope** as I find ctags inadequate for large code bases.

Good work.

**Comment by ludovicchabant** [comment]
I don't know GNU Global but I have an experimental version of Cscope support kinda working, but not awesome. I'm very new to Cscope so I don't know exactly how it works besides the most simple stuff... I'll push it here so people can comment on it but it's not there yet.

For example, I'm not sure how to incrementally update a Cscope database when you save a file?

**Comment by badeip** [comment]
*For example, I'm not sure how to incrementally update a Cscope database when you save a file?*

I haven't tried that, and non-incremental update of the database might be too slow for lage projects
Async *neovim* integration would probably be a very popular future addition for your plugin though.

In the script below, I generate the database files in a separate directory, and move them to the correct destination when populated, to minimize downtime.

I wrote the following **Cscope** and **GNU Global** script for generating tags, and *grep.vim* for vim integration. Feel free to rip whatever you might find useful.

*tags.sh*:

    #! /bin/bash
    
    DBDIR=.db
    if [ $# -ne 1 ]; then
        SRCDIR=~/src/master
    else
        SRCDIR=$1
    fi
    
    populate_file_list()
    {
        printf "[+] Populating file list...\n"
        if [ -d .git ]; then
            git ls-tree -r --name-only master | egrep ".c$|.cpp$|.h$|.hpp$" | egrep -v "boost|testtools" > $DBDIR/cscope.files
        else
            find $SRCDIR -type f \( -name \*.c -o -name \*.cpp -o -name \*.cc -o -name \*.hpp -o -name \*.h -o -name \*.asm -o -name \*.def \) | egrep -sv 'boost' > $DBDIR/cscope.files
        fi
    
    }
    
    build_global()
    {
        printf "[+] Building GNU Global database...\n"
        gtags -f $DBDIR/cscope.files $DBDIR
        cp -pf GPATH GRTAGS GTAGS $DBDIR 2>/dev/null
    }
    
    build_ctags()
    {
        ctags -L $DBDIR/cscope.files
    }
    
    build_cscope()
    {
        printf "[+] Quoting file paths...\n"
    #    sed -i 's/^/"/;s/$/"/' $DBDIR/cscope.files
    
        printf "[+] Building cscope database...\n"
        cscope -vbq -i $DBDIR/cscope.files -f $DBDIR/cscope.out
    
    }
    
    main()
    {
        cd $SRCDIR
        mkdir -p $DBDIR
    
        printf "[+] Copying existing databases...\n"
    
        populate_file_list
        build_global
    
        build_cscope
    
        printf "[+] Moving global database files...\n"
        mv -f $DBDIR/{GTAGS,GPATH,GRTAGS} .
    
        printf "[+] Moving cscope database files...\n"
        mv -f $DBDIR/{cscope.out,cscope.out.in,cscope.out.po} .
    }
    
    
    main $*



*grep.vim*:

    silent com! -nargs=+ Tag call CMD_tag(<f-args>)
    function! CMD_gtag(...)
        let l:needle=""
        if a:0 == 1
            let l:needle .= expand("<cword>")
        else
            let l:needle .= a:2
        endif
    
        echom "GTAG"
        if a:1 == "c"            " callers
            set grepprg =global\ -rx\ $*
        elseif a:1 == "g"        " implementation
            set grepprg =global\ -dx\ $*
        elseif a:1 == "s"        " symbol
            set grepprg =global\ -sx\ $*
        elseif a:1 == "f"       " path
            set grepprg =global\ -Pi\ $*
        els
            set grepprg =global\ -dx\ $*
        endif
    
        let grepformat = &grepformat
        let grepprg = &grepprg
        set grepformat=%*[^\ ]%*[\ ]%l%*[\ ]%f%*[\ ]%m
    
        let cmd = 'grep ' . l:needle
        exe cmd | copen | redraw!
    
        let &grepprg = grepprg
        let &grepformat = grepformat
    endfunction
    
    function! CMD_ctag(...)
        let l:needle=""
        if a:0 == 1
            let l:needle .= expand("<cword>")
        else
            let l:needle .= a:2
        endif
    
        echom "CTAG"
        let l:cmd="cscope -qRL -0 "
        if a:1 == "c"            " callers
            let l:cmd ="cscope -qRL -3 " . l:needle
        elseif a:1 == "g"        " implementation
            let l:cmd ="cscope -qRL -1 " . l:needle
        elseif a:1 == "s"        " symbol
            let l:cmd ="cscope -qRL -0 " . l:needle
        elseif a:1 == "f"       " path
            let l:cmd ="cscope -qRL -8 " . l:needle
        else
            let l:cmd ="cscope -qRL -0 " . l:needle
        endif
    
        let cmd_output=system(l:cmd)
        if cmd_output == ""
            echohl WarningMsg | echomsg "Error: " . l:needle . " not found" | echohl None
            return
        endif
    
        let tmpfile = tempname()
        let curfile = expand("%")
    
        exe "redir! > " . tmpfile
        if curfile != ""
            silent echon curfile . " dummy " . line(".") . " " . getline(".") . "\n"
        endif
        silent echon cmd_output
        redir END
    
        let old_efm = &efm
        set efm=%f\ %*[^\ ]\ %l\ %m
    
        exe "silent! cfile " . tmpfile
        let &efm = old_efm
    
        botright copen
        call delete(tmpfile)
    endfunction
    
    fu! CMD_tag(...)
        " prefer cscope
        if filereadable("cscope.out")
            silent call call("CMD_ctag", a:000)
        else
            silent call call("CMD_gtag", a:000)
        endif
    endfunction

**Comment by ludovicchabant** [comment]
So I had to fight `hg-git` a bit but I got the changes pushed. Super early prototype cscope support is available with 90c86c745e237277bfec1d9d47160dc5db74c153 but note that the script is only on Window for now (my Mac/Linux projects are too small so I haven't ported it yet). You enable it by setting `g:gutentags_modules` to `['ctags', 'cscope']`.

What's the `-v` option you pass to cscope? I can't find it on the man page?
Either way, if there is no way to re-index a sub-set of the cscope database when the user saves a file, I'm not sure cscope support is really useful. Best case, Gutentags could take care of re-indexing when you open a new project, and providing an easy command to re-index manually, but that's far from the ideal user experience that I'm shooting for with this plugin... :(
